subsys/net/l2/ethernet/dsa/dsa.c

```diff
@@ -0,0 +1,209 @@
+/** @file
+ * @brief DSA related functions
+ */
+
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_dsa, CONFIG_NET_DSA_LOG_LEVEL);
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include <net/net_core.h>
+#include <net/ethernet.h>
+#include <net/net_mgmt.h>
+#include <net/dsa.h>
+
+/*
+ * Helper functions
+ */
+struct dsa_context *dsa_get_context(struct net_if *iface)
+{
+	struct ethernet_context *eth_ctx;
+
+	eth_ctx = net_if_l2_data(iface);
+	if (eth_ctx == NULL) {
+		LOG_ERR("Iface %p context not available!", iface);
+		return NULL;
+	}
+
+	return eth_ctx->dsa_ctx;
+}
+
+/*
+ * Store, in the ethernet_context for master interface, the original
+ * eth_tx() function, which will send packet with tag appended.
+ */
+int dsa_register_master_tx(struct net_if *iface, dsa_send_t fn)
+{
+	struct ethernet_context *ctx = net_if_l2_data(iface);
+
+	ctx->dsa_send = fn;
+	return 0;
+}
+
+#ifdef CONFIG_NET_L2_ETHERNET
+bool dsa_is_port_master(struct net_if *iface)
+{
+	/* First check if iface points to ETH interface */
+	if (net_if_l2(iface) == &NET_L2_GET_NAME(ETHERNET)) {
+		/* Check its capabilities */
+		if (net_eth_get_hw_capabilities(iface) &
+		    ETHERNET_DSA_MASTER_PORT) {
+			return true;
+		}
+	}
+
+	return false;
+}
+#else
+bool dsa_is_port_master(struct net_if *iface)
+{
+	return false;
+}
+#endif
+
+/*
+ * RECEIVE HANDLING CODE - ingress (ETH -> DSA slave ports)
+ */
+
+static int dsa_check_iface(struct net_if *iface)
+{
+	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+		return -ENOENT;
+	}
+
+	if (!((net_eth_get_hw_capabilities(iface) & ETHERNET_DSA_MASTER_PORT) ||
+	    (net_eth_get_hw_capabilities(iface) & ETHERNET_DSA_SLAVE_PORT))) {
+		return -ESRCH;
+	}
+
+	return 0;
+}
+
+
+int dsa_register_recv_callback(struct net_if *iface, dsa_net_recv_cb_t cb)
+{
+	struct ethernet_context *ctx;
+	int ret;
+
+	ret = dsa_check_iface(iface);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (cb) {
+		ctx = net_if_l2_data(iface);
+		ctx->dsa_recv_cb = cb;
+	}
+
+	return 0;
+}
+
+static int dsa_net_recv(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct ethernet_context *ctx;
+	int ret;
+
+	ret = dsa_check_iface(iface);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ctx = net_if_l2_data(iface);
+	if (ctx->dsa_recv_cb) {
+		return ctx->dsa_recv_cb(iface, pkt);
+	}
+
+	return 0;
+}
+
+struct net_if *dsa_recv_set_iface(struct net_if *iface, struct net_pkt **pkt)
```

Comment:
I would just merge the content of that function into dsa_net_recv().

Now that you got rid of using dsa_net_revc from net_recv_data, I think you can just call dsa_net_recv in relevant drivers directy (with the iface and the **pkt).

It's seems more logical to do it that way.

I think it's my last comment. rest looks good to me

Reply:
Please correct me if I've misunderstood you:
You propose to:
1. Rename `dsa_recv_set_iface` (line 126 above) to `dsa_net_recv`
2. The content of `static int dsa_net_recv(struct net_if *iface, struct net_pkt *pkt)` (line 108) merge into the renamed function from point 1 above.

Am I correct?

Reply:
I've updated the code - please check